url1a41b8e9f2_mr1azl_INF345_tgz-pJ8-TestParserJ8/src/xx/MonCompiler.java

```diff
@@ -0,0 +1,2234 @@
+						// "and")[1].getSteps());
+						predResult.clear();
+						predResult.add(n);
+						Object[] newArgs = getNewAndArgs(node, predResult, "or");
+
+						boolean test = Utils.asBoolean(or(newArgs));
+						if (test && !tmp.contains(n)) {
+							tmp.add(n);
+						}
+
+					}
+					if (((NodeTest) node).getFunctionCalls().containsKey(
+							"" + FUNCTION_LAST)) {
+						if (cur == last) {
+							tmp.add(n);
+						}
+					}
+					if (((NodeTest) node).getFunctionCalls().containsKey(
+							"" + FUNCTION_CONTAINS)) {
+						NodeTest[] argument = ((NodeTest) node)
+								.getFunctionCalls().get("" + FUNCTION_CONTAINS);
+						predResult = (List<Node>) locationPath(predResult,
+								((NodeTest) argument[0]).getSteps());
+						for (Node nx : predResult) {
+							String keyword = ((NodeTest) argument[1])
+									.getNodeValue();
+							if (contains(nx, keyword)) {
+								tmp.add(n);
+							}
+						}
+					}
+					if (((NodeTest) node).getFunctionCalls().containsKey(
+							"" + FUNCTION_NOT)) {
+						NodeTest[] argument = ((NodeTest) node)
+								.getFunctionCalls().get("" + FUNCTION_NOT);
+						predResult = (List<Node>) locationPath(predResult,
+								((NodeTest) argument[0]).getSteps());
+						for (Node nx : predResult) {
+							String keyword = ((NodeTest) argument[1])
+									.getNodeValue();
+							if (contains(nx, keyword)) {
+								tmp.add(n);
+							}
+						}
+					}
+					if (((NodeTest) node).getFunctionCalls().containsKey(
+							"" + FUNCTION_STARTS_WITH)) {
+						NodeTest[] argument = ((NodeTest) node)
+								.getFunctionCalls().get(
+										"" + FUNCTION_STARTS_WITH);
+						predResult = (List<Node>) locationPath(predResult,
+								((NodeTest) argument[0]).getSteps());
+						for (Node nx : predResult) {
+							String keyword = ((NodeTest) argument[1])
+									.getNodeValue();
+							if (startsWith(nx, keyword)) {
+								tmp.add(n);
+							}
+						}
+					}
+					if (((NodeTest) node).getFunctionCalls().containsKey(
+							"" + FUNCTION_LANG)) {
+						NodeTest[] argument = ((NodeTest) node)
+								.getFunctionCalls().get("" + FUNCTION_LANG);
+
+						String keyword = ((NodeTest) argument[0])
+								.getNodeValue();
+						if (lang(n, keyword)) {
+							tmp.add(n);
+
+						}
+					}
+					if (((NodeTest) node).getFunctionCalls().containsKey(
+							"" + FUNCTION_SUBSTRING)) {
+
+						NodeTest[] argument = ((NodeTest) node)
+								.getFunctionCalls()
+								.get("" + FUNCTION_SUBSTRING);
+						predResult = (List<Node>) locationPath(predResult,
+								((NodeTest) argument[0]).getSteps());
+						for (Node nx : predResult) {
+							int i1 = Integer.parseInt(((NodeTest) argument[1])
+									.getNodeValue());
+							int i2 = Integer.parseInt(((NodeTest) argument[2])
+									.getNodeValue());
+							NodeTest ret = new NodeTest();
+							ret.setNodeType(NODE_TYPE_STRING);
+							ret.setNodeValue(subString(nx, i1, i2));
+
+							tmp.add(ret);
+
+						}
+					}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 1132:1152, 1493:1512.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.